### PR TITLE
Add trainer for DQN

### PR DESCRIPTION
**Pull request by NoB0**
This PR includes:

- An abstract `Trainer` class
- A subclass `TrainerDQN` to train a dialogue policy using DQN algorithm
- A script to run training
- Configuration files
- Utils

Note that some methods related to user simulation will be part of another PR.

### CODE REVIEW

File: reinforcement_learning/config/train_a2c_policy.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Participants' configuration
+usersim_config: data/config/usersim_agenda_config.yaml
+agent_config: data/config/moviebot_config_rb_nlu.yaml
+
+# Model's parameters
+model:
+  model_type: a2c
```
Comment (WerLaj) [inline reinforcement_learning/config/train_a2c_policy.yaml:7]: Can we also call it algorithm for the consistency between config files?

File: reinforcement_learning/config/train_dqn_policy.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# Model's parameters
+model:
+  algorithm: dqn
+  policy_input_size: 8
+  hidden_size: 128
+  model_path: models/policy/dqn_policy.pt
+
+# Training hyperparameters
+hyperparams:
+  num_episodes: 1000
```
Comment (WerLaj) [inline reinforcement_learning/config/train_dqn_policy.yaml:14]: Nit: n_episodes for the consistency between config files?

File: reinforcement_learning/train_dialogue_policy.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    dqn_hyperparameters_group.add_argument(
+        "--num_episodes",
+        type=int,
+        help="Number of episodes.",
+        dest="hyperparams.num_episodes",
+    )
+    dqn_hyperparameters_group.add_argument(
+        "--tau",
+        type=float,
+        help="Tau.",
```
Comment (WerLaj) [inline reinforcement_learning/train_dialogue_policy.py:125]: Could you make it a little bit more descriptive?

File: reinforcement_learning/train_dialogue_policy.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    a2c_hyperparameters_group.add_argument(
+        "--a2c-gamma",
+        type=float,
+        help="Discount factor.",
+        dest="hyperparams.gamma",
+    )
+    a2c_hyperparameters_group.add_argument(
+        "--lambda",
+        type=float,
+        help="Lambda.",
```
Comment (WerLaj) [inline reinforcement_learning/train_dialogue_policy.py:170]: Could it be a bit more descriptive?

File: reinforcement_learning/training/trainer_dqn.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+Transition = namedtuple(
+    "Transition", ("state", "action", "next_state", "reward")
+)
+
+
+class ReplayMemory(object):
+    def __init__(self, capacity: int) -> None:
+        """Create Replay memory.
+        Args:
```
Comment (WerLaj) [inline reinforcement_learning/training/trainer_dqn.py:28]: Nit: New line missing

File: reinforcement_learning/training/trainer_dqn.py [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+
+        # Save policy
+        self.save_model()
+
+        # Report to wandb
+        self.report_to_wandb(num_episodes)
+        wandb.finish()
+
+    def report_to_wandb(self, num_episodes: int) -> None:
+        """Add training plots and data to wandb summary."""
```
Comment (WerLaj) [inline reinforcement_learning/training/trainer_dqn.py:285]: Nit: Adds

File: reinforcement_learning/utils.py [hunk 7]
```diff
@@ -15,4 +39,122 @@ def build_agenda_based_simulator(
...
+    ]
+    actions += [
+        DialogueAct(
+            AgentIntents.RECOMMEND,
+        ),
+        DialogueAct(
+            AgentIntents.CONTINUE_RECOMMENDATION,
+        ),
+    ]
+    actions.append(DialogueAct(AgentIntents.INFORM))
```
Comment (WerLaj) [inline reinforcement_learning/utils.py:127]: Could you do it for all or at least some of these intents at once? It seems to be an easy function but it still takes a lot of lines of code.

**Review by WerLaj — APPROVED** [review]
LGTM with some minor comments

File: reinforcement_learning/config/train_a2c_policy.yaml (hunk 1 above)
Comment (NoB0) [reply reinforcement_learning/config/train_a2c_policy.yaml:7]: Good point! Renamed.

File: reinforcement_learning/training/trainer_dqn.py (hunk 6 above)
Comment (NoB0) [reply reinforcement_learning/training/trainer_dqn.py:285]: Done.

File: reinforcement_learning/training/trainer_dqn.py (hunk 5 above)
Comment (NoB0) [reply reinforcement_learning/training/trainer_dqn.py:28]: Done.

File: reinforcement_learning/utils.py (hunk 7 above)
Comment (NoB0) [reply reinforcement_learning/utils.py:127]: Done.

File: reinforcement_learning/train_dialogue_policy.py (hunk 3 above)
Comment (NoB0) [reply reinforcement_learning/train_dialogue_policy.py:125]: Done.

File: reinforcement_learning/train_dialogue_policy.py (hunk 4 above)
Comment (NoB0) [reply reinforcement_learning/train_dialogue_policy.py:170]: Done.

**Pull request merged**